nautobot/extras/migrations/0057_collect_roles_from_related_apps_roles.py

```diff
@@ -61,22 +61,46 @@ def bulk_create_roles(apps, roles_to_create, content_types):
     """Bulk create role and set its contenttypes"""
     Role = apps.get_model("extras", "Role")
     ContentType = apps.get_model("contenttypes", "ContentType")
+    ObjectChange = apps.get_model("extras", "ObjectChange")
+
+    old_role_ct = None
+    new_role_ct = None
+
+    if hasattr(roles_to_create, "model"):
+        old_role_ct = ContentType.objects.get_for_model(roles_to_create.model)
+        # No sense fetching this if we don't have an old role content type
+        new_role_ct = ContentType.objects.get_for_model(Role)
 
     # Exclude roles whose names are already exists from bulk create.
     existing_roles = Role.objects.values_list("name", flat=True)
     roles_instances = [
-        Role(
-            name=data.name,
-            slug=data.slug,
-            description=data.description,
-            color=getattr(data, "color", color_map["default"]),
-            weight=getattr(data, "weight", None),
+        (
+            data,
+            Role(
+                name=data.name,
+                slug=data.slug,
+                description=data.description,
+                color=getattr(data, "color", color_map["default"]),
+                weight=getattr(data, "weight", None),
+            ),
         )
         for data in roles_to_create
         if data.name not in existing_roles
     ]
     if roles_instances:
-        Role.objects.bulk_create(roles_instances, batch_size=1000)
+        for old_role, new_role in roles_instances:
+            # TODO: We need to handle new role collisions
+            new_role.save()
+            if old_role_ct:
+                # Move over existing object change records to the new role we created
+                ObjectChange.objects.filter(changed_object_type=old_role_ct, changed_object_id=old_role.pk).update(
+                    changed_object_type=new_role_ct, changed_object_id=new_role.pk
+                )
```

Comment:
Won't this skip over ObjectChanges for roles that happen to already have been created with a different content-type?

Reply:
Unless my logic is extremely faulty the way I've read through the order of these migrations is that won't ever happen. The migration doesn't appear to collapse Roles into a single object, hence the collision problem that still exists.

I would ultimately like to refactor the way these migrations happen before we ship 2.0.

Currently, we have a dependency on the non-extras apps schema migrations on this data migration and we loosely associate the old role name to the new ones by hoping they have a similar name and/or slug.

I've tested this pretty extensively locally, restoring the DBs and re-running for different scenarios.